components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/EngineProcessPriorityMiddleware.kt

```diff
@@ -0,0 +1,40 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.state.engine.middleware
+
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.EngineAction
+import mozilla.components.browser.state.selector.findTab
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.concept.engine.EngineSession
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * [Middleware] implementation responsible for calling [EngineSession.updateSessionPriority] on
+ * [EngineSession] instances.
+ */
+internal class EngineProcessPriorityMiddleware : Middleware<BrowserState, BrowserAction> {
+    private val logger = Logger("EngineProcessPriorityMiddleware")
+
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        when (action) {
+            is EngineAction.UpdateEngineSessionPriorityAction -> {
```

Comment:
What we'd want here is to set the priority to high for the active session, and default for the non-active session. So with that in mind, we need to know when a new session is selected and the previous one un-selected.

In the BrowserStore, we have the action `LinkEngineSessionAction` which is when a new tab is created which is what we should observe. This is better than knowing the selected tab, because the "linking" is done when a new tab is created right before it is selected (so via the context menu, selecting a tab from the tray, or after restoring the selected tab). 

If we observe the `UnlinkEngineSessionAction`, we know when to do the opposite of the above.

Similar to the `WebExtensionMiddleware`, we can find the linked engineSession and directly call `updateSessionPriority` on it instead of dispatching another action to do it.